flowauth/backend/tests/test_commands.py

```diff
@@ -64,16 +64,73 @@ def test_init_db_force(app):
             add_admin_command, ["DUMMY_ADMINISTRATOR", "DUMMY_ADMINISTATOR_PASSWORD"]
         )
         assert len(User.query.all()) > 0
+        app.config["DB_IS_SET_UP"].clear()
         result = runner.invoke(init_db_command, ["--force"])
         assert len(User.query.all()) == 0
 
 
+def test_demo_data_only_sets_up_once(app, caplog):
+    """
+    Demo data should only be created once.
+    """
+    with app.app_context():
+        runner = app.test_cli_runner()
+        app.config["DB_IS_SET_UP"].clear()
+        result = runner.invoke(demodata)
+        result = runner.invoke(demodata)
+        assert len(User.query.all()) == 2
+        assert len(Group.query.all()) == 3
+    assert "Database already set up by another worker, skipping." in caplog.text
+
+
+def test_demo_data_skips_when_in_progress(app, caplog):
+    """
+    Shouldn't attempt to create demo data if something else is.
+    """
+    with app.app_context():
+        runner = app.test_cli_runner()
+        app.config["DB_IS_SET_UP"].clear()
+        app.config["DB_IS_SETTING_UP"].set()
```

Comment:
Technically the use of `DB_IS_SETTING_UP` as a flag is an internal implementation detail, but I can't think of an easy way to reliably trigger the test condition in a different way so I guess it's fine to use it in the test too.

Reply:
Actually that's a good point - these should probably be on `g` instead really. I'll change it.